### ISSUE TITLE
toStartOf...  functions do not work with datetime64 extended range

### ISSUE THREAD

**Issue by Misaka10091** [ClickHouse/ClickHouse#25284]
**How to reproduce**
* ClickHouse server: 21.4.7.3
* Queries
    ```clickhouse
    WITH toDateTime64('1940-01-01 01:00:00', 0) AS dt64
    SELECT dt64,toStartOfDay(dt64)
    ```
* Results
  ```text
  ┌────────────────dt64─┬──toStartOfDay(dt64)─┐
  │ 1940-01-01 01:00:00 │ 2076-02-06 06:28:16 │
  └─────────────────────┴─────────────────────┘
  ```
**Expected behavior**
return date `1940-01-01 00:00:00` instead of `2076-02-06 06:28:16`

**Additional context**
most date functions return unexpected value
this seems to overflowed by date before 1970

**Comment by zvonand** [comment]
https://github.com/ClickHouse/ClickHouse/issues/22928#issuecomment-816600831

Not supposed to work

**Comment by amosbird** [comment]
We can either 1. make `toStartOfDay` and variants aware of `Date32` and `DateTime64` and promote the return type, which will break some old use cases; 2. introduce `toStartOfDay64` variants.  @alexey-milovidov WDYT

**Comment by alexey-milovidov** [comment]
I think (1) is better.

**Comment by amosbird** [comment]
We cannot modify the return type of toStartOfDay because it might be used in table's partition by clause. We already saw similar issue with the modulo function, and came up with a compromise solution https://github.com/ClickHouse/ClickHouse/pull/24157 . But it will complicate index analysis and some issues are inevitable https://github.com/ClickHouse/ClickHouse/issues/28800 . As a result, I'd like to propose a different solution based on auto-AST transformation. We add a new function toStartOfDay64 which returns DateTime64 if argument is DateTime64. We rewrite AST with toStartOfDay64 function when encountering toStartOfDay for queries only. Newly created table will have partition by toStartOfDay64 stored. Old table definition with partition by toStartOfDay will still work. @alexey-milovidov

**Comment by amosbird** [comment]
Actually we can also fix the `modulo` issue using the same idea.

**Comment by alexey-milovidov** [comment]
That's good idea, let's do it.

**Comment by amosbird** [comment]
We still need a way to resolve backward compatibility when running both new and old instances.

**Comment by alexey-milovidov** [comment]
```
milovidov-desktop :) SET enable_extended_results_for_datetime_functions = 1

SET enable_extended_results_for_datetime_functions = 1

Query id: 10e4ed1e-8d22-484c-9378-bc74a51e808b

Ok.

0 rows in set. Elapsed: 0.000 sec. 

milovidov-desktop :) WITH toDateTime64('1940-01-01 01:00:00', 0) AS dt64
                     SELECT dt64,toStartOfDay(dt64)

WITH toDateTime64('1940-01-01 01:00:00', 0) AS dt64
SELECT
    dt64,
    toStartOfDay(dt64)

Query id: 55782084-4f85-4cc2-97fa-cd8cbb01a4a8

┌────────────────dt64─┬──toStartOfDay(dt64)─┐
│ 1940-01-01 01:00:00 │ 1940-01-01 00:00:00 │
└─────────────────────┴─────────────────────┘

1 row in set. Elapsed: 0.001 sec.
```

**Comment by alexey-milovidov** [comment]
Fixed by #41910.

**Comment by rschu1ze** [comment]
It is planned that "enable_extended_results_for_datetime_functions" becomes the default after 1. stabilization and 2. all other date time functions w/o extended precision support were fixed.
